2020 Kentucky State House/Senate Districts

Kara77779 opened the pull request:
# 2020 Kentucky State House/Senate Districts

## Redistricting requirements
In Kentucky, we consult [NCSL Redistricting Law 2020](https://documents.ncsl.org/wwwncsl/Redistricting-Census/Redistricting-Law-2020_NCSL%20FINAL.pdf) and impose the following constraints. In our simulations, legislative districts must:

1. Be contiguous [NCSL, 184]
2. Equal population [NCSL, 23]
3. While compactness is commonly considered traditional criterion, it is not a statutory requirement in Kentucky [NCSL, 184]

### Algorithmic Constraints
We enforce a maximum population deviation of 5.0%.

## Data Sources
Data for Kentucky comes from the ALARM Project's [2020 Redistricting Data Files](https://alarm-redist.github.io/posts/2021-08-10-census-2020/).

## Pre-processing Notes
No manual pre-processing decisions were necessary.

## Simulation Notes
We sample 10,000 districting plans for Kentucky's lower house across 5 independent runs of the SMC algorithm.
No special techniques were needed to produce the sample.

We sample 10,000 districting plans for Kentucky's upper house across 5 independent runs of the SMC algorithm.
No special techniques were needed to produce the sample.


## Validation

## SHD
<img width="3000" height="5400" alt="validation_20260110_1808" src="https://github.com/user-attachments/assets/b05240a2-922b-48d7-a738-376988fca2d1" />

```
SMC with Merge-Split MCMC Steps: 10,000 sampled plans of 100 districts on 3,693 units
Plans sampled on Linking Edge Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:                       
• `weight_type` = optimal             
• `seq_alpha` = 1                     
• `pop_temper` = 0                    
Mergesplit Parameters:                
• `total_ms_steps` = 99               
• `mh_accept_per_smc` = 44            
• `pair_rule` = uniform               
Plan diversity 80% range: 0.72 to 0.80
82 rhat values are `NA`or KY_shd_2020

Largest R-hat values for ordered summary statistics:
        adv_16         adv_20         arv_16         arv_20      comp_edge    comp_polsby  county_splits 
         1.026          1.014          1.006          1.019          1.007          1.017          1.005 
         e_dem          e_dvs           egap    muni_splits        ndshare            ndv            nrv 
         1.006          1.033          1.012          1.002          1.037          1.019          1.014 
         pbias       plan_dev       pop_aian      pop_asian      pop_black       pop_hisp       pop_nhpi 
         1.021          1.002          1.027          1.023          1.024          1.021          1.016 
     pop_other    pop_overlap        pop_two      pop_white         pr_dem pre_16_dem_cli pre_16_rep_tru 
         1.043          1.007          1.019          1.009          1.010          1.022          1.010 
pre_20_dem_bid pre_20_rep_tru      total_vap uss_16_dem_gra uss_16_rep_pau uss_20_dem_mcg uss_20_rep_mcc 
         1.017          1.019          1.003          1.018          1.008          1.013          1.018 
      vap_aian      vap_asian      vap_black       vap_hisp       vap_nhpi      vap_other        vap_two 
         1.025          1.022          1.025          1.024          1.015          1.031          1.012 
     vap_white 
         1.009 
• R-hat ≤ 1.05: 4218                  
• 1.05 < R-hat ≤ 1.1: 0               
• R-hat > 1.1: 0                      
• Total R-hats: 4218                  
Sampling diagnostics for SMC run 1 of 5 (10,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1       919 (45.9%)    100.0%        1.20 1,275 (101%) 
Split 2     1,080 (54.0%)     99.5%        1.19   991 ( 78%) 
Split 3     1,105 (55.3%)     98.7%        1.09 1,031 ( 82%) 
Split 4     1,222 (61.1%)     98.1%        0.99 1,040 ( 82%) 
Split 5     1,200 (60.0%)     97.0%        0.95 1,054 ( 83%) 
Split 6     1,302 (65.1%)     96.3%        0.87 1,083 ( 86%) 
Split 7     1,346 (67.3%)     95.1%        0.83 1,104 ( 87%) 
Split 8     1,402 (70.1%)     93.7%        0.76 1,115 ( 88%) 
Split 9     1,395 (69.7%)     93.2%        0.76 1,126 ( 89%) 
Split 10    1,434 (71.7%)     93.0%        0.71 1,134 ( 90%) 
Split 11    1,475 (73.7%)     89.5%        0.68 1,135 ( 90%) 
Split 12    1,528 (76.4%)     89.8%        0.62 1,119 ( 89%) 
Split 13    1,529 (76.4%)     90.0%        0.63 1,166 ( 92%) 
Split 14    1,588 (79.4%)     88.5%        0.57 1,133 ( 90%) 
Split 15    1,615 (80.7%)     88.0%        0.55 1,184 ( 94%) 
Split 16    1,606 (80.3%)     87.9%        0.55 1,175 ( 93%) 
Split 17    1,642 (82.1%)     86.5%        0.51 1,194 ( 94%) 
Split 18    1,669 (83.5%)     85.4%        0.50 1,175 ( 93%) 
Split 19    1,655 (82.7%)     85.3%        0.50 1,171 ( 93%) 
Split 20    1,693 (84.6%)     83.5%        0.46 1,202 ( 95%) 
Split 21    1,696 (84.8%)     83.1%        0.46 1,183 ( 94%) 
Split 22    1,708 (85.4%)     82.1%        0.44 1,202 ( 95%) 
Split 23    1,732 (86.6%)     80.2%        0.42 1,206 ( 95%) 
Split 24    1,722 (86.1%)     79.7%        0.43 1,213 ( 96%) 
Split 25    1,761 (88.1%)     78.4%        0.40 1,193 ( 94%) 
Split 26    1,760 (88.0%)     77.8%        0.40 1,201 ( 95%) 
Split 27    1,763 (88.1%)     77.5%        0.39 1,189 ( 94%) 
Split 28    1,768 (88.4%)     76.8%        0.39 1,220 ( 97%) 
Split 29    1,772 (88.6%)     75.0%        0.38 1,220 ( 97%) 
Split 30    1,777 (88.8%)     74.8%        0.37 1,233 ( 98%) 
Split 31    1,783 (89.2%)     73.7%        0.36 1,225 ( 97%) 
Split 32    1,796 (89.8%)     71.8%        0.37 1,212 ( 96%) 
Split 33    1,814 (90.7%)     71.8%        0.35 1,228 ( 97%) 
Split 34    1,806 (90.3%)     70.8%        0.35 1,220 ( 97%) 
Split 35    1,828 (91.4%)     71.9%        0.33 1,232 ( 97%) 
Split 36    1,830 (91.5%)     69.2%        0.32 1,222 ( 97%) 
Split 37    1,844 (92.2%)     69.1%        0.31 1,212 ( 96%) 
Split 38    1,824 (91.2%)     66.5%        0.31 1,231 ( 97%) 
Split 39    1,839 (91.9%)     66.3%        0.30 1,222 ( 97%) 
Split 40    1,844 (92.2%)     66.1%        0.30 1,227 ( 97%) 
Split 41    1,847 (92.3%)     64.1%        0.30 1,212 ( 96%) 
Split 42    1,850 (92.5%)     64.8%        0.30 1,224 ( 97%) 
Split 43    1,865 (93.3%)     62.8%        0.28 1,261 (100%) 
Split 44    1,871 (93.6%)     62.7%        0.27 1,244 ( 98%) 
Split 45    1,868 (93.4%)     61.1%        0.27 1,236 ( 98%) 
Split 46    1,874 (93.7%)     61.0%        0.27 1,243 ( 98%) 
Split 47    1,879 (93.9%)     60.2%        0.27 1,245 ( 98%) 
Split 48    1,878 (93.9%)     59.5%        0.26 1,249 ( 99%) 
Split 49    1,876 (93.8%)     60.1%        0.27 1,234 ( 98%) 
Split 50    1,884 (94.2%)     58.7%        0.26 1,249 ( 99%) 
Split 51    1,887 (94.3%)     57.0%        0.25 1,222 ( 97%) 
Split 52    1,894 (94.7%)     56.6%        0.25 1,256 ( 99%) 
Split 53    1,895 (94.7%)     57.4%        0.24 1,249 ( 99%) 
Split 54    1,901 (95.1%)     55.5%        0.24 1,228 ( 97%) 
Split 55    1,903 (95.2%)     55.4%        0.23 1,234 ( 98%) 
Split 56    1,905 (95.3%)     54.0%        0.23 1,228 ( 97%) 
Split 57    1,906 (95.3%)     52.6%        0.23 1,232 ( 97%) 
Split 58    1,909 (95.5%)     54.2%        0.22 1,237 ( 98%) 
Split 59    1,913 (95.7%)     51.1%        0.22 1,250 ( 99%) 
Split 60    1,914 (95.7%)     52.1%        0.22 1,250 ( 99%) 
Split 61    1,914 (95.7%)     51.9%        0.22 1,252 ( 99%) 
Split 62    1,920 (96.0%)     50.0%        0.21 1,238 ( 98%) 
Split 63    1,926 (96.3%)     51.7%        0.20 1,218 ( 96%) 
Split 64    1,923 (96.2%)     50.4%        0.20 1,260 (100%) 
Split 65    1,922 (96.1%)     50.5%        0.20 1,248 ( 99%) 
Split 66    1,925 (96.2%)     47.8%        0.20 1,245 ( 98%) 
Split 67    1,930 (96.5%)     46.5%        0.19 1,254 ( 99%) 
Split 68    1,933 (96.6%)     46.8%        0.19 1,244 ( 98%) 
Split 69    1,929 (96.4%)     44.7%        0.20 1,272 (101%) 
Split 70    1,932 (96.6%)     44.9%        0.19 1,240 ( 98%) 
Split 71    1,934 (96.7%)     44.8%        0.19 1,234 ( 98%) 
Split 72    1,941 (97.0%)     44.1%        0.18 1,256 ( 99%) 
Split 73    1,941 (97.1%)     43.4%        0.18 1,248 ( 99%) 
Split 74    1,942 (97.1%)     42.6%        0.17 1,261 (100%) 
Split 75    1,943 (97.2%)     42.6%        0.17 1,270 (100%) 
Split 76    1,943 (97.2%)     41.8%        0.17 1,267 (100%) 
Split 77    1,944 (97.2%)     41.5%        0.17 1,241 ( 98%) 
Split 78    1,945 (97.2%)     42.1%        0.17 1,254 ( 99%) 
Split 79    1,946 (97.3%)     40.5%        0.17 1,258 (100%) 
Split 80    1,949 (97.4%)     40.3%        0.16 1,241 ( 98%) 
Split 81    1,950 (97.5%)     40.9%        0.16 1,259 (100%) 
Split 82    1,949 (97.5%)     39.0%        0.16 1,224 ( 97%) 
Split 83    1,955 (97.7%)     38.2%        0.15 1,238 ( 98%) 
Split 84    1,956 (97.8%)     38.5%        0.15 1,237 ( 98%) 
Split 85    1,953 (97.6%)     36.0%        0.16 1,244 ( 98%) 
Split 86    1,956 (97.8%)     38.0%        0.15 1,235 ( 98%) 
Split 87    1,957 (97.9%)     37.5%        0.15 1,249 ( 99%) 
Split 88    1,958 (97.9%)     36.9%        0.15 1,238 ( 98%) 
Split 89    1,958 (97.9%)     35.8%        0.15 1,228 ( 97%) 
Split 90    1,960 (98.0%)     35.1%        0.15 1,217 ( 96%) 
Split 91    1,959 (98.0%)     34.3%        0.15 1,242 ( 98%) 
Split 92    1,963 (98.2%)     32.8%        0.14 1,221 ( 97%) 
Split 93    1,962 (98.1%)     33.1%        0.14 1,225 ( 97%) 
Split 94    1,966 (98.3%)     32.6%        0.13 1,206 ( 95%) 
Split 95    1,969 (98.4%)     31.8%        0.13 1,245 ( 98%) 
Split 96    1,971 (98.6%)     32.6%        0.12 1,218 ( 96%) 
Split 97    1,971 (98.6%)     30.2%        0.12 1,196 ( 95%) 
Split 98    1,976 (98.8%)     31.7%        0.11 1,191 ( 94%) 
Split 99    1,979 (98.9%)     30.4%        0.10 1,113 ( 88%) 
Resample    1,979 (98.9%)       NA%          NA 1,895 (150%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (10,000 samples)
           Acc. rate MS Steps per Plan
MS Step 1      51.6%                44
MS Step 2      50.5%                88
MS Step 3      50.3%                88
MS Step 4      50.4%                88
MS Step 5      50.1%                88
MS Step 6      49.9%                88
MS Step 7      50.0%               132
MS Step 8      49.9%                88
MS Step 9      49.8%               132
MS Step 10     49.7%               132
MS Step 11     49.5%               132
MS Step 12     49.2%               132
MS Step 13     49.1%               132
MS Step 14     49.0%               132
MS Step 15     48.6%               132
MS Step 16     48.6%               132
MS Step 17     48.6%               132
MS Step 18     48.3%               132
MS Step 19     48.2%               132
MS Step 20     48.2%               132
MS Step 21     48.0%               132
MS Step 22     47.8%               132
MS Step 23     47.7%               132
MS Step 24     47.6%               132
MS Step 25     47.3%               132
MS Step 26     47.2%               132
MS Step 27     46.8%               132
MS Step 28     46.9%               132
MS Step 29     46.4%               132
MS Step 30     46.3%               132
MS Step 31     46.1%               132
MS Step 32     46.0%               132
MS Step 33     45.8%               132
MS Step 34     45.5%               132
MS Step 35     45.4%               132
MS Step 36     45.4%               132
MS Step 37     45.2%               132
MS Step 38     44.7%               132
MS Step 39     44.2%               132
MS Step 40     44.5%               132
MS Step 41     44.2%               132
MS Step 42     44.0%               132
MS Step 43     43.5%               132
MS Step 44     43.5%               132
MS Step 45     43.3%               132
MS Step 46     43.1%               132
MS Step 47     42.8%               132
MS Step 48     42.5%               132
MS Step 49     42.2%               132
MS Step 50     42.1%               132
MS Step 51     41.9%               132
MS Step 52     41.6%               132
MS Step 53     41.5%               132
MS Step 54     41.1%               132
MS Step 55     41.0%               132
MS Step 56     40.7%               132
MS Step 57     40.8%               132
MS Step 58     40.4%               132
MS Step 59     40.1%               132
MS Step 60     39.9%               132
MS Step 61     39.7%               132
MS Step 62     39.6%               132
MS Step 63     39.3%               132
MS Step 64     39.0%               132
MS Step 65     38.8%               132
MS Step 66     38.5%               132
MS Step 67     38.6%               132
MS Step 68     38.3%               132
MS Step 69     38.0%               132
MS Step 70     37.6%               132
MS Step 71     37.5%               132
MS Step 72     37.3%               132
MS Step 73     37.0%               132
MS Step 74     36.6%               132
MS Step 75     36.6%               132
MS Step 76     36.3%               132
MS Step 77     36.3%               132
MS Step 78     35.9%               132
MS Step 79     35.6%               132
MS Step 80     35.5%               132
MS Step 81     35.2%               132
MS Step 82     34.9%               132
MS Step 83     34.7%               132
MS Step 84     34.5%               132
MS Step 85     34.3%               132
MS Step 86     34.0%               132
MS Step 87     33.6%               132
MS Step 88     33.3%               132
MS Step 89     33.2%               132
MS Step 90     32.9%               176
MS Step 91     32.7%               176
MS Step 92     32.3%               176
MS Step 93     32.3%               176
MS Step 94     32.0%               176
MS Step 95     31.9%               176
MS Step 96     31.5%               176
MS Step 97     31.3%               176
MS Step 98     30.7%               176
MS Step 99     30.6%               176

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log
weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary statistics should be
between 1 and 1.05.
```

## SSD
<img width="3000" height="5400" alt="validation_20260110_1810" src="https://github.com/user-attachments/assets/f182589a-5523-475e-a4af-50f5d7567117" />

```
SMC with Merge-Split MCMC Steps: 10,000 sampled plans of 38 districts on 3,693 units
Plans sampled on Linking Edge Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0
Mergesplit Parameters:
• `total_ms_steps` = 37
• `mh_accept_per_smc` = 13
• `pair_rule` = uniform
Plan diversity 80% range: 0.72 to 0.85
62 rhat values are `NA`
Largest R-hat values for ordered summary statistics:
        adv_16         adv_20         arv_16         arv_20      comp_edge    comp_polsby  county_splits        e_dem.x        e_dem.y        e_dvs.x        e_dvs.y         egap.x         egap.y    muni_splits 
         1.015          1.011          1.008          1.009          1.006          1.009          1.010          1.016          1.016          1.023          1.023          1.013          1.013          1.014 
       ndshare            ndv            nrv        pbias.x        pbias.y       plan_dev       pop_aian      pop_asian      pop_black       pop_hisp       pop_nhpi      pop_other    pop_overlap        pop_two 
         1.022          1.013          1.010          1.005          1.005          1.000          1.019          1.027          1.017          1.010          1.016          1.020          1.038          1.017 
     pop_white       pr_dem.x       pr_dem.y pre_16_dem_cli pre_16_rep_tru pre_20_dem_bid pre_20_rep_tru      total_vap uss_16_dem_gra uss_16_rep_pau uss_20_dem_mcg uss_20_rep_mcc       vap_aian      vap_asian 
         1.012          1.015          1.015          1.015          1.006          1.011          1.010          1.005          1.015          1.007          1.011          1.009          1.014          1.031 
     vap_black       vap_hisp       vap_nhpi      vap_other        vap_two      vap_white 
         1.018          1.015          1.022          1.021          1.013          1.008 
• R-hat ≤ 1.05: 1762
• 1.05 < R-hat ≤ 1.1: 0
• R-hat > 1.1: 0
• Total R-hats: 1762
Sampling diagnostics for SMC run 1 of 5 (10,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1       892 (44.6%)    100.0%        1.13 1,270 (100%) 
Split 2     1,022 (51.1%)     98.7%        1.15   974 ( 77%) 
Split 3     1,139 (56.9%)     95.9%        1.02   989 ( 78%) 
Split 4     1,204 (60.2%)     93.9%        0.98 1,041 ( 82%) 
Split 5     1,282 (64.1%)     91.4%        0.88 1,066 ( 84%) 
Split 6     1,327 (66.3%)     89.2%        0.81 1,107 ( 88%) 
Split 7     1,381 (69.1%)     87.1%        0.77 1,093 ( 86%) 
Split 8     1,442 (72.1%)     83.3%        0.72 1,121 ( 89%) 
Split 9     1,469 (73.4%)     82.3%        0.68 1,149 ( 91%) 
Split 10    1,479 (73.9%)     79.6%        0.67 1,159 ( 92%) 
Split 11    1,501 (75.1%)     77.6%        0.63 1,147 ( 91%) 
Split 12    1,526 (76.3%)     75.6%        0.61 1,147 ( 91%) 
Split 13    1,581 (79.0%)     73.1%        0.56 1,160 ( 92%) 
Split 14    1,591 (79.5%)     70.1%        0.55 1,167 ( 92%) 
Split 15    1,614 (80.7%)     67.1%        0.52 1,168 ( 92%) 
Split 16    1,618 (80.9%)     65.7%        0.50 1,188 ( 94%) 
Split 17    1,660 (83.0%)     62.7%        0.47 1,176 ( 93%) 
Split 18    1,679 (84.0%)     64.0%        0.46 1,189 ( 94%) 
Split 19    1,684 (84.2%)     60.1%        0.45 1,220 ( 97%) 
Split 20    1,709 (85.5%)     59.8%        0.43 1,178 ( 93%) 
Split 21    1,722 (86.1%)     58.3%        0.42 1,183 ( 94%) 
Split 22    1,729 (86.5%)     55.5%        0.40 1,204 ( 95%) 
Split 23    1,750 (87.5%)     54.0%        0.39 1,209 ( 96%) 
Split 24    1,759 (88.0%)     54.4%        0.38 1,212 ( 96%) 
Split 25    1,786 (89.3%)     49.7%        0.35 1,193 ( 94%) 
Split 26    1,801 (90.1%)     49.7%        0.33 1,211 ( 96%) 
Split 27    1,799 (90.0%)     49.2%        0.34 1,232 ( 97%) 
Split 28    1,797 (89.8%)     47.2%        0.33 1,188 ( 94%) 
Split 29    1,806 (90.3%)     45.7%        0.33 1,219 ( 96%) 
Split 30    1,831 (91.6%)     43.5%        0.31 1,199 ( 95%) 
Split 31    1,839 (91.9%)     43.5%        0.30 1,225 ( 97%) 
Split 32    1,855 (92.7%)     42.4%        0.28 1,201 ( 95%) 
Split 33    1,873 (93.6%)     42.5%        0.27 1,179 ( 93%) 
Split 34    1,883 (94.1%)     37.8%        0.25 1,199 ( 95%) 
Split 35    1,893 (94.6%)     36.7%        0.24 1,191 ( 94%) 
Split 36    1,910 (95.5%)     36.7%        0.22 1,169 ( 92%) 
Split 37    1,915 (95.7%)     34.8%        0.22 1,164 ( 92%) 
Resample    1,915 (95.7%)       NA%          NA 1,829 (145%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (10,000 samples)
           Acc. rate MS Steps per Plan
MS Step 1      54.3%                13
MS Step 2      51.6%                26
MS Step 3      51.0%                26
MS Step 4      50.6%                26
MS Step 5      50.3%                26
MS Step 6      49.5%                26
MS Step 7      48.1%                39
MS Step 8      47.7%                39
MS Step 9      47.4%                39
MS Step 10     46.9%                39
MS Step 11     45.9%                39
MS Step 12     45.3%                39
MS Step 13     44.2%                39
MS Step 14     43.9%                39
MS Step 15     43.1%                39
MS Step 16     42.2%                39
MS Step 17     41.6%                39
MS Step 18     40.4%                39
MS Step 19     40.1%                39
MS Step 20     39.4%                39
MS Step 21     38.3%                39
MS Step 22     37.9%                39
MS Step 23     37.1%                39
MS Step 24     36.5%                39
MS Step 25     36.3%                39
MS Step 26     35.5%                39
MS Step 27     34.8%                39
MS Step 28     34.5%                39
MS Step 29     34.2%                39
MS Step 30     33.1%                39
MS Step 31     32.5%                52
MS Step 32     31.9%                52
MS Step 33     31.2%                52
MS Step 34     31.1%                52
MS Step 35     30.4%                52
MS Step 36     29.6%                52
MS Step 37     29.3%                52

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary statistics should be
between 1 and 1.05.

```

## Checklist

- [x] I have followed the [instructions](https://github.com/alarm-redist/fifty-states/blob/main/CONTRIBUTING.md)
- [x] I have updated the [tracker](https://docs.google.com/spreadsheets/d/1nV1vrzQZBCil9MdQV_8AOQr4Y4X2evYO9R4WeZkiO3g/edit?usp=sharing)
- [x] All `TODO` lines from the template code have been removed
- [x] I have merged in the main branch and then recalculated summary statistics
- [x] I have run `enforce_style()` to format my code
- [x] The documentation copied above is up-to-date 
- [x] There are no data files in this pull request
- [x] None of the file output paths (for the `redist_map` and `redist_plans` objects, and summary statistics) have been edited

@philipwosull

Kara77779 committed:
Add IN-leg-2020

Kara77779 committed:
Merge branch 'main' of https://github.com/alarm-redist/fifty-states into IN-leg-2020

Kara77779 committed:
Add Kentucky legislative redistricting analysis

Kara77779 committed:
Run enforce_style() to format code

philipwosull commented:
@Kara77779 Any updates on this? I would recommend playing around with the size of the psuedo-county and adding a total splits constraint and seeing if that can fix the splits issues

Kara77779 commented:
> > That looks much better, total county splits looks great. Could you try adding a splits constraint on the county as well (`add_constr_plan_splits` function). Lets see if we can drive those county splits down as well.
> 
> Hi Philip, I added add_constr_plan_splits(strength = 2) along with the existing add_constr_total_splits(strength = 3). Results: County splits: distribution now ~5-8, enacted (~5) at left edge ✓ Total county splits: distribution ~14-20, enacted (~21) now at right edge The county splits improved, but the total county splits shifted - enacted moved from center to right edge. Should I adjust the strength values to balance both metrics, or does this look acceptable? <img alt="validation_20260303_0855" width="2000" height="2000" src="https://private-user-images.githubusercontent.com/149135632/557530991-111c72c2-b111-4533-a84a-a39ccf98e73c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzM4ODM1NTEsIm5iZiI6MTc3Mzg4MzI1MSwicGF0aCI6Ii8xNDkxMzU2MzIvNTU3NTMwOTkxLTExMWM3MmMyLWIxMTEtNDUzMy1hODRhLWEzOWNjZjk4ZTczYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMxOVQwMTIwNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iOWY2ZDRkMWU3NzNkNThjMmE4ZGRhNzQxNzNjMzNjZGFkNTUxMjA4MzMwMmNmYjc3NmI3NDZkYzViY2MzYzgzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.H9wZDItC0hj4fKvwUtCdIGJlRaK_F1aIN-BSVZR0uwA">



> Hi Philip,
> I added add_constr_plan_splits(strength = 2) along with the existing add_constr_total_splits(strength = 3). Results:
> County splits: distribution now ~5-8, enacted (~5) at left edge ✓
> Total county splits: distribution ~14-20, enacted (~21) now at right edge
> The county splits improved, but the total county splits shifted - enacted moved from center to right edge. Should I adjust the strength values to balance both metrics, or does this look acceptable?

Hi @christopherkenny @philipwosull
I ran the full simulation for KY Senate with:

add_constr_total_splits(strength = 3, admin = county)
add_constr_plan_splits(strength = 2, admin = county)
nsims = 2100, runs = 5

The splits metrics look good:

Total county splits: R-hat = 1.018 
County splits: enacted (~5) within distribution 

However, many other R-hat values are above 1.05, and the summary shows "Chains have not converged."
What would you recommend to improve convergence?

SMC with Merge-Split MCMC Steps: 10,000 sampled plans of 38 districts on 3,693 units
Plans sampled on Linking Edge Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• weight_type = optimal
• seq_alpha = 1
• pop_temper = 0
Mergesplit Parameters:
• total_ms_steps = 37
• mh_accept_per_smc = 33
• pair_rule = uniform
Plan diversity 80% range: 0.55 to 0.73
31 rhat values are NA
Largest R-hat values for ordered summary statistics:
adv_16 adv_20 arv_16 arv_20 comp_bbox_reock
❌1.26 ❌1.42 ❌1.433 ❌1.371 ❌1.127
comp_edge comp_polsby county_splits e_dem e_dvs
❌1.056 ❌1.109 ❌1.067 ❌1.384 ❌1.292
egap muni_splits ndshare ndv nrv
❌1.363 1.023 ❌1.29 ❌1.372 ❌1.47
pbias plan_dev pop_aian pop_asian pop_black
❌1.141 ❌1.051 ❌1.284 ❌1.481 ❌1.293
pop_hisp pop_nhpi pop_other pop_overlap pop_two
❌1.32 ❌1.391 ❌1.39 ❌1.107 ❌1.343
pop_white pr_dem pre_16_dem_cli pre_16_rep_tru pre_20_dem_bid
❌1.519 ❌1.429 ❌1.252 ❌1.39 ❌1.326
pre_20_rep_tru total_county_splits total_muni_splits total_vap uss_16_dem_gra
❌1.345 1.018 ❌1.062 ❌1.243 ❌1.154
uss_16_rep_pau uss_20_dem_mcg uss_20_rep_mcc vap_aian vap_asian
❌1.155 ❌1.331 ❌1.385 ❌1.171 ❌1.535
vap_black vap_hisp vap_nhpi vap_other vap_two
❌1.255 ❌1.293 ❌1.398 ❌1.6 ❌1.432
vap_white
❌1.505
• R-hat ≤ 1.05: 502
• 1.05 < R-hat ≤ 1.1: 550
• R-hat > 1.1: 665
• Total R-hats: 1717
✖ WARNING: Chains have not converged.
Sampling diagnostics for SMC run 1 of 5 (10,000 samples)
Eff. samples (%) Acc. rate Log wgt. sd Max. unique
Split 1 469 (22.3%) 100.0% 1.8 1,341 (101%)
Split 2 211 (10.0%) 98.6% 2.7 779 ( 59%)
Split 3 201 (9.6%) 96.0% 2.7 437 ( 33%)
Split 4 208 (9.9%) 94.8% 2.8 443 ( 33%)
Split 5 196 (9.3%) 92.5% 2.6 442 ( 33%)
Split 6 193 (9.2%) 88.9% 2.8 446 ( 34%)
Split 7 173 (8.2%) 86.6% 2.7 450 ( 34%)
Split 8 202 (9.6%) 84.7% 2.7 486 ( 37%)
Split 9 180 (8.6%) 81.1% 2.9 483 ( 36%)
Split 10 208 (9.9%) 80.5% 2.7 469 ( 35%)
Split 11 249 (11.9%) 77.2% 2.7 557 ( 42%)
Split 12 248 (11.8%) 74.8% 2.9 556 ( 42%)
Split 13 323 (15.4%) 72.1% 3.1 558 ( 42%)
Split 14 325 (15.5%) 71.1% 3.0 597 ( 45%)
Split 15 387 (18.4%) 68.4% 2.9 625 ( 47%)
Split 16 424 (20.2%) 63.9% 3.0 683 ( 51%)
Split 17 462 (22.0%) 62.8% 3.2 684 ( 52%)
Split 18 507 (24.2%) 60.0% 2.9 687 ( 52%)
Split 19 448 (21.3%) 58.8% 3.0 737 ( 56%)
Split 20 501 (23.9%) 56.7% 2.8 722 ( 54%)
Split 21 466 (22.2%) 55.6% 2.9 750 ( 56%)
Split 22 483 (23.0%) 55.2% 3.0 723 ( 54%)
Split 23 440 (21.0%) 54.3% 2.8 730 ( 55%)
Split 24 469 (22.3%) 50.9% 2.8 723 ( 54%)
Split 25 480 (22.8%) 48.1% 2.6 751 ( 57%)
Split 26 449 (21.4%) 46.3% 2.6 770 ( 58%)
Split 27 472 (22.5%) 45.3% 2.6 779 ( 59%)
Split 28 479 (22.8%) 44.0% 2.5 769 ( 58%)
Split 29 514 (24.5%) 43.8% 2.7 774 ( 58%)
Split 30 564 (26.8%) 41.8% 2.6 803 ( 60%)
Split 31 616 (29.4%) 41.7% 2.6 810 ( 61%)
Split 32 640 (30.5%) 37.0% 2.5 857 ( 65%)
Split 33 644 (30.7%) 35.3% 2.5 875 ( 66%)
Split 34 736 (35.1%) 31.0% 2.3 913 ( 69%)
Split 35 861 (41.0%) 29.3% 1.9 979 ( 74%)
Split 36 1,101 (52.4%) 27.5% 1.7 990 ( 75%)
Split 37 1,181 (56.3%) 22.9% 1.0 1,019 ( 77%)
Resample 1,181 (56.3%) NA% NA 1,524 (115%)

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (10,000 samples)
Acc. rate MS Steps per Plan
MS Step 1 35.5% 33
MS Step 2 24.7% 99
MS Step 3 20.2% 165
MS Step 4 18.5% 165
MS Step 5 17.4% 198
MS Step 6 16.4% 198
MS Step 7 16.1% 231
MS Step 8 15.6% 231
MS Step 9 15.6% 231
MS Step 10 15.2% 231
MS Step 11 15.7% 231
MS Step 12 15.9% 231
MS Step 13 16.4% 231
MS Step 14 17.2% 231
MS Step 15 18.0% 198
MS Step 16 18.6% 198
MS Step 17 19.2% 198
MS Step 18 19.6% 198
MS Step 19 19.8% 198
MS Step 20 19.8% 198
MS Step 21 19.7% 198
MS Step 22 19.8% 198
MS Step 23 19.8% 198
MS Step 24 19.8% 198
MS Step 25 19.9% 198
MS Step 26 19.6% 198
MS Step 27 19.8% 198
MS Step 28 19.9% 198
MS Step 29 19.8% 198
MS Step 30 19.8% 198
MS Step 31 20.1% 198
MS Step 32 20.5% 165
MS Step 33 20.9% 165
MS Step 34 21.0% 165
MS Step 35 20.9% 165
MS Step 36 21.1% 165
MS Step 37 21.4% 165

• Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log
weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary statistics should be
between 1 and 1.05.
• SMC convergence: Increase the number of samples. If you are experiencing low plan diversity or bottlenecks as
well, address those issues first.

<img width="3000" height="5400" alt="image" src="https://github.com/user-attachments/assets/3e5e7174-f535-4341-8034-1d26b38a0d3d" />
